Build against qdldl 0.1.7 and run tests

adament opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [X] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [X] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

adament commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found some lint.

Here's what I've got...


For **recipe/meta.yaml**:

 * This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.

adament commented:
@conda-forge-admin, please rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

traversaro commented:
Thanks!

h-vetinari commented on recipe/meta.yaml:
```diff
@@ -24,13 +24,23 @@ requirements:
     - ninja
   host:
     - libqdldl ={{ qdldlversion }}
+    - numpy # For unittests
+    - scipy =1.11.4 # For unittests, as of 0.6.3 imports randn from scipy
 
 test:
```
Please move these to the test section. The `*` is to avoid generating several builds for several numpy/python versions.
```suggestion

test:
  requires:
    - numpy *
    # 0.6.3 imports randn from scipy
    - scipy <1.12
```

adament replied:
Unfortunately they are used in the build process to generate the unit test code. I have added numpy to ignore_run_exports, do I need to do more to ensure they stay as build only dependencies?

adament replied:
I just realised that given numpy and scipy is used to generate code during build time but not something the package is built against, I guess they should be `build` dependencies not `host` dependencies?

h-vetinari commented on recipe/meta.yaml:
```diff
@@ -24,13 +24,23 @@ requirements:
+    - build/out/osqp_demo*
+    - build/out/osqp_tester*
   commands:
+    - build\out\osqp_demo.exe  # [win]
+    - build\out\osqp_tester.exe  # [win]
+    - ./build/out/osqp_demo  # [not win]
+    - ./build/out/osqp_tester  # [not win]
```
Please move this to the end of the test section. Tests should go in increasing order of complexity - first simple file existence checks (if those fail, more complicated ones make no sense), then more integrated workflows.

adament replied:
That makes sense, thank you.

h-vetinari commented on recipe/meta.yaml:
```diff
@@ -24,13 +24,23 @@ requirements:
+    - scipy =1.11.4 # For unittests, as of 0.6.3 imports randn from scipy
 
 test:
+  source_files:
+    - build/out/osqp_demo*
+    - build/out/osqp_tester*
```
These files weren't included in the build, and thus haven't been made relocatable. You'd need to point to `$PREFIX/lib/libosqp.so` instead of the one in `$SRC_DIR/build/...`, e.g. using something like patchelf (linux) / install_name_tool (osx).

Or you install those demo files directly to `$PREFIX/...` during installation; the result is essentially the same, because everything under `test.source_files:` gets packaged in the final artefact anyway (to be able to run tests locally)

adament replied:
I have installed them in `$PREFIX` as you suggested, I have just used cmake default `install()` behaviour which means they are installed in the `$PREFIX/bin` folder, is there a more suitable place to put them?

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found some lint.

Here's what I've got...


For **recipe/meta.yaml**:

 * The requirements section contained an unexpected subsection name. ignore_run_exports is not a valid subsection name.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

adament commented:
@conda-forge-admin, please rerender

adament commented:
I am very sorry for the commit spam, I had trouble getting a local build working.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you but ran into some issues. Please check the output logs of the GitHub actions workflow below for more details. You can also ping conda-forge/core for further assistance or you can try [rerendering locally](https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-with-conda-smithy-locally).


The following suggestions might help debug any issues:
* Is the `recipe/{meta.yaml,recipe.yaml}` file valid?
* If there is a `recipe/conda-build-config.yaml` file in the feedstock make sure that it is compatible with the current [global pinnnings](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml).
* Is the fork used for this PR on an organization or user GitHub account? Automated rerendering via the webservices admin bot only works for user GitHub accounts.


<sub>This message was generated by GitHub actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/11415306093](https://github.com/conda-forge/conda-forge-webservices/actions/runs/11415306093).</sub>

adament committed:
The exit commands on windows made the test shell exit

This had the effect that only the first test was ever run on Windows.

Co-authored-by: h-vetinari <h.vetinari@gmx.com>

adament commented:
@h-vetinari, @traversaro I think this is ready for another review. I am not sure what the issue with Travis is. The error does not make a lot of sense to me:
```
Adding in variants from internal_defaults
Adding in variants from /home/conda/feedstock_root/.ci_support/linux_aarch64_.yaml
Attempting to finalize metadata for libosqp
Reloading output folder: ...working... done
Getting pinned dependencies: ...working... failed
Reloading output folder: ...working... done
Getting pinned dependencies: ...working... failed
Reloading output folder: ...working... done
Getting pinned dependencies: ...working... failed
Reloading output folder: ...working... done
Getting pinned dependencies: ...working... failed
```

traversaro commented:
Travis is really problematic as now. Could you switch the arm and PPC builds to cross-compilation so you can use azure?

adament committed:
Cross-compile linux powerpc and arm

adament commented:
Thank you very much, I have tried to do that. However it just dawned on me, that it is concerning that the tests work when trying to cross-compile? If the package is cross-compiled how come the host machine is able to execute the demo and unittest applications which are supposedly compiled for another architecture?

traversaro commented:
> Thank you very much, I have tried to do that. However it just dawned on me, that it is concerning that the tests work when trying to cross-compile? If the package is cross-compiled how come the host machine is able to execute the demo and unittest applications which are supposedly compiled for another architecture?

Tests and in general the build system uses qemu to run in emulation the executables compiled for the target architecture (except for osx-arm64 were there is no emulator available).

adament commented:
In that case I hope things are ready now.

h-vetinari commented on recipe/meta.yaml:
```diff
@@ -35,10 +45,12 @@ test:
-    - if exist $PREFIX$\\Library\\lib\\osqp.lib (exit 0) else (exit 1)  # [win]
-    - if exist $PREFIX$\\Library\\bin\\osqp.dll (exit 0) else (exit 1)  # [win]
-    - if exist %PREFIX%\\Library\\lib\\cmake\\osqp\\osqp-config.cmake (exit 0) else (exit 1)  # [win]
+    - if exist %PREFIX%\\Library\\include\\osqp\\osqp.h (echo.>nul) else (mkdir . 2>nul)  # [win]
+    - if exist %PREFIX%\\Library\\lib\\osqp.lib (echo.>nul) else (mkdir . 2>nul)  # [win]
+    - if exist %PREFIX%\\Library\\bin\\osqp.dll (echo.>nul) else (mkdir . 2>nul)  # [win]
+    - if exist %PREFIX%\\Library\\lib\\cmake\\osqp\\osqp-config.cmake (echo.>nul) else (mkdir . 2>nul)  # [win]
```
This is not suitable. Here's a suggestion that also shortens the lines a bit.
```suggestion
    - if not exist %LIBRARY_INC%\osqp\osqp.h exit 1                     # [win]
    - if not exist %LIBRARY_LIB%\osqp.lib exit 1                        # [win]
    - if not exist %LIBRARY_BIN%\osqp.dll exit 1                        # [win]
    - if not exist %LIBRARY_LIB%\cmake\osqp\osqp-config.cmake exit 1    # [win]

```

h-vetinari commented on recipe/meta.yaml:
```diff
@@ -13,19 +13,29 @@ source:
+    # Unittest source code is generated at build time using numpy and scipy
+    - python
+    - numpy
+    - scipy =1.11.4 # as of 0.6.3 imports randn from scipy
```
Don't pin to patch version. The constraint should say what it means to do, which is: avoid too-new scipy because it removed a deprecated feature that libosqp still relies on
```suggestion
    # v0.6.3 imports scipy.randn which was removed in 1.12
    - scipy <1.12
```

h-vetinari commented on recipe/meta.yaml:
```diff
@@ -13,19 +13,29 @@ source:
     - patches/0001-Add-OSQP_RESPECT_BUILD_SHARED_LIBS-option.patch
     # use libqdldl from conda-forge, build interface in osqp directly
     - patches/0002-build-qdldl-interface-as-part-of-osqp.patch
+      # install osqp_demo and osqp_tester executables to easily make the relocatable,
+      # see: https://github.com/conda-forge/libosqp-feedstock/pull/12#discussion_r1807117589
```
```suggestion
      # install osqp_demo and osqp_tester executables to easily make them relocatable,
      # see: https://github.com/conda-forge/libosqp-feedstock/pull/12#discussion_r1807117589
```

h-vetinari commented:
> I am very sorry for the commit spam, I had trouble getting a local build working.

It would be helpful to clean up the commit history a bit (squash commits that belong together or correct each other; remove all the rerender and only have a single one at the end). Are you familiar with `git rebase -i origin/main`? If not I can help you do that.

adament committed:
Build against qdldl 0.1.7

adament committed:
Build unittests and install osqp_demo and osqp_tester

And test against the bundled demo application and the unittests. The
osqp_demo and osqp_tester applications are installed in PREFIX as that
easily makes them relocatable, and they would be bundled in the package
anyway to support local testing.

Co-authored-by: h-vetinari <h.vetinari@gmx.com>

adament committed:
MNT: Re-rendered with conda-build 24.9.0, conda-smithy 3.43.0, and conda-forge-pinning 2024.10.19.11.12.22

adament commented:
> > I am very sorry for the commit spam, I had trouble getting a local build working.
> 
> It would be helpful to clean up the commit history a bit (squash commits that belong together or correct each other; remove all the rerender and only have a single one at the end). Are you familiar with `git rebase -i origin/main`? If not I can help you do that.

I tried to clean up the history, I hope this is somewhat like what you had in mind and I did not do it too aggressively?

h-vetinari reviewed (approved):
The cleaned-up history is great, thanks a lot! 👏 

(you didn't need to give me co-authorship for minor review, but thanks)
